### DISCUSSION TITLE
Spatie translatable in combination with array casted attributes

### DISCUSSION THREAD

**Discussion by Senexis** [filamentphp/filament#11374] [Help]
### Package

Form builder

### Package Version

v3.2.28

### How can we help you?

Hey all!

I'm currently trying to get a model attribute which is also cast to `array` to be translatable within form builder.

This is my current setup minimalized. I deliberately added the full trait qualifier, since there's multiple `Translatable` traits and to show which ones I'm using where.

```php
class MyModel extends Model
{
    use HasFactory, HasTranslations, HasUuids;

    protected $casts     = ['data' => 'array'];
    public $translatable = ['data'];

    // migration: $table->json('data')
}
```

```php
class EditMyModel extends EditRecord
{
    use [Filament\Resources\Pages\EditRecord\Concerns\]Translatable;

    protected static string $resource = MyModelResource::class;

    protected function getHeaderActions(): array
    {
        return [
            Actions\LocaleSwitcher::make(),
            Actions\DeleteAction::make(),
        ];
    }
}
```

```php
class MyModelResource extends Resource
{
    use [Filament\Resources\Concerns\]Translatable;

    protected static ?string $model          = MyModelModel::class;
    protected static ?string $navigationIcon = 'heroicon-o-rectangle-stack';

    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Components\TextInput::make('data.myfield')->label('My field')->required(),
                // more fields...
            ]);
    }

    // table...
}
```

Submitting data in the `data.myfield` field saves correctly for the default loaded locale. It can also be retrieved correctly, and the database result is `{"en": {"myfield": "My value"}}` as expected.

The problem starts when switching to another locale. When I do that the `My field` value is emptied, as expected. When I then go ahead and fill the field and save it, the field will report a "My field is required" error. Checking out the payload of the save request, we find the `data` field being completely empty.

Removing the required attribute from `My field` results in an incorrect `{"en": {"myfield": null}}` result even when the field is filled.

Is anyone familiar with this issue? If so, is there any known method of making it so that array fields in other languages are set correctly?

**Comment by Senexis** [comment]
**Update:** As far as I can tell, this is due to the `$record->getTranslation()` ignoring casts, and always returning an empty string (`""`) instead of an array for a language that has not been filled yet. Thus, it seems this is a Spatie issue.

I've managed to find a (admittedly not very generic) fix for this in the form of a mutator on the page:

```php
class EditMyModel extends EditRecord
{
    use Translatable;

    protected static string $resource = MyModelResource::class;

    protected function getHeaderActions(): array
    {
        return [
            Actions\LocaleSwitcher::make(),
            Actions\DeleteAction::make(),
        ];
    }

    /**
     * @param  array<string, mixed>  $data
     * @return array<string, mixed>
     */
    protected function mutateFormDataBeforeFill(array $data): array
    {
        if (!is_array($data['data'])) {
            $data['data'] = [];
        }

        return $data;
    }
}
```

If possible I would love to use the Model's casts attributes and do this dynamically, but for now this seems fine. If anybody has a suggestion or improvement to this, please do reach out and let me know!

**Reply by mathias-aa** [reply]
This solution worked for me ! thx
